csharp/src/ParallelSelectExt.cs

```diff
@@ -0,0 +1,368 @@
+
+  /// <summary>
+  ///   Options for ParallelSelect and ParallelWait.
+  ///   Parallelism is limited to the number of threads.
+  /// </summary>
+  /// <param name="cancellationToken">Cancellation token used for stopping the enumeration</param>
+  public ParallelTaskOptions(CancellationToken cancellationToken)
+    : this(0,
+           cancellationToken)
+  {
+  }
+}
+
+[PublicAPI]
+public static class ParallelSelectExt
+{
+  /// <summary>
+  ///   Iterates over the input enumerable and spawn multiple parallel tasks.
+  ///   The maximum number of tasks in flight at any given moment is given in the `parallelTaskOptions`.
+  ///   All results are collected in-order.
+  /// </summary>
+  /// <param name="enumerable">Enumerable to iterate on</param>
+  /// <param name="parallelTaskOptions">Options (eg: parallelismLimit, cancellationToken)</param>
+  /// <typeparam name="T">Return type of the tasks</typeparam>
+  /// <returns>Asynchronous results of tasks</returns>
+  [PublicAPI]
+  public static async IAsyncEnumerable<T> ParallelWait<T>(this IEnumerable<Task<T>> enumerable,
+                                                          ParallelTaskOptions       parallelTaskOptions = default)
+  {
+    var parallelism       = parallelTaskOptions.ParallelismLimit;
+    var cancellationToken = parallelTaskOptions.CancellationToken;
+
+    // CancellationTokenSource used to cancel all tasks inflight upon errors
+    using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+
+    // Queue of tasks
+    var queue = new Queue<Task<T>>();
+    // Semaphore to limit the parallelism
+    // Semaphore is created with one resource already acquired
+    using var sem = new SemaphoreSlim(parallelism - 1,
+                                      parallelism);
+
+    // Iterate over the enumerable
+    foreach (var x in enumerable)
+    {
+      // Prepare the new semaphore acquisition
+      var semAcquire = sem.WaitAsync(cts.Token);
+
+      // Async closure that waits for the task x and releases the semaphore
+      async Task<T> TaskLambda()
+      {
+        var res = await x.ConfigureAwait(false);
+        sem.Release();
+        return res;
+      }
+
+      // We can enqueue the task
+      queue.Enqueue(Task.Run(TaskLambda,
+                             cts.Token));
+
+      // Dequeue tasks as long as the semaphore is not acquired yet
+      while (true)
+      {
+        var front = queue.Count != 0
+                      ? queue.Peek()
+                      : null;
+        // Select the first task to be ready
+        // There is no need to explicitly wait on cancellation: semAcquire natively supports cancellation
+        using var which = await Task.WhenAny(front ?? semAcquire,
+                                             semAcquire)
+                                    .ConfigureAwait(false);
+
+        // If there is an error or cancellation has been requested, we must throw the exception
+        which.ThrowIfError();
```

Comment:
No need.
Exception is triggered line 159.

Reply:
Only if the task is `front`, but in case it is `semAcquire`, this is required. It could be done in the if, though. But I just like it there. I think it is easier to understand.